presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizationManager.java

```diff
@@ -0,0 +1,268 @@
+            }
+        }, startDelay, discoveryInterval.toMillis(), MILLISECONDS);
+    }
+
+    private long getStartDelay()
+    {
+        Long secondsAgo = shardOrganizerDao.getShardOrganizerLastRunSecondsAgo(currentNodeIdentifier);
+        if (secondsAgo == null) {
+            shardOrganizerDao.insertNode(currentNodeIdentifier);
+            secondsAgo = shardOrganizerDao.getShardOrganizerLastRunSecondsAgo(currentNodeIdentifier);
+        }
+        return max(discoveryInterval.toMillis() - SECONDS.toMillis(secondsAgo), 0);
+    }
+
+    @VisibleForTesting
+    void discoverShardsAndSubmitJobs()
+    {
+        Collection<Long> enabledTableIds = metadataDao.getOrganizationEnabledTables();
+        Set<ShardMetadata> allShards = shardManager.getNodeShards(currentNodeIdentifier);
+        ListMultimap<Long, ShardMetadata> tableShards = Multimaps.index(allShards, ShardMetadata::getTableId);
+
+        for (Map.Entry<Long, Collection<ShardMetadata>> entry : tableShards.asMap().entrySet()) {
+            long tableId = entry.getKey();
+
+            // If not enabled or not sorted, nothing to organize, skip the table
+            if (!enabledTableIds.contains(tableId) || metadataDao.listSortColumns(tableId).isEmpty()) {
+                continue;
+            }
+
+            Table tableInfo = metadataDao.getTableInformation(tableId);
+            Set<ShardMetadata> filteredShards = entry.getValue().stream()
+                    .filter(shard -> !organizer.inProgress(shard.getShardUuid()))
+                    .collect(toSet());
+
+            Collection<ShardIndexInfo> indexInfos = toShardIndexInfo(dbi, metadataDao, tableInfo, filteredShards, true);
+            Set<OrganizationSet> organizationSets = createOrganizationSets(tableInfo, indexInfos);
+
+            log.info("Created %s organization set(s) from %s shard(s) for table ID %s", organizationSets.size(), filteredShards.size(), tableId);
+            organizationSets.forEach(organizer::enqueue);
+        }
+    }
+
+    @VisibleForTesting
+    static Set<OrganizationSet> createOrganizationSets(Table tableInfo, Collection<ShardIndexInfo> shards)
+    {
+        Collection<Collection<ShardIndexInfo>> partitionedShards = getShardsByDaysBuckets(tableInfo, shards);
+        ImmutableSet.Builder<OrganizationSet> builder = ImmutableSet.builder();
+        for (Collection<ShardIndexInfo> indexInfos : partitionedShards) {
+            builder.addAll(getOverlappingOrganizationSets(tableInfo, indexInfos));
+        }
+        return builder.build();
+    }
+
+    private static Set<OrganizationSet> getOverlappingOrganizationSets(Table tableInfo, Collection<ShardIndexInfo> shards)
+    {
+        if (shards.size() <= 1) {
+            return ImmutableSet.of();
+        }
+
+        boolean allShardsHaveRange = shards.stream()
+                .map(ShardIndexInfo::getSortRange)
+                .allMatch(Optional::isPresent);
+
+        if (!allShardsHaveRange) {
```

Comment:
This is just a sanity check. Actually we can get rid of this as it will never happen (because sort columns are set in the columns table, an shardrange is simply a min max on those columns).